File: apps/api/internal/distribution/mastodon.go
package distribution

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
	"unicode/utf8"
)

// Mastodon posta chamadas sociais numa instância Mastodon via REST API.
type Mastodon struct {
	instance string
	token    string
	http     *http.Client
}

// NewMastodon cria um adapter Mastodon para a instância e token fornecidos.
func NewMastodon(instanceURL, token string) *Mastodon {
	return &Mastodon{
		instance: strings.TrimRight(instanceURL, "/"),
		token:    token,
		http:     &http.Client{Timeout: 30 * time.Second},
	}
}

func (m *Mastodon) Platform() string { return "mastodon" }
func (m *Mastodon) Kind() Kind       { return KindSocial }

// Publish posta um status no Mastodon. Rejeita textos > 500 runes.
// Retorna a URL pública do toot.
func (m *Mastodon) Publish(ctx context.Context, p Payload) (string, error) {
	if utf8.RuneCountInString(p.Text) > 500 {
		return "", fmt.Errorf("mastodon: texto excede 500 caracteres")
	}
	buf, _ := json.Marshal(map[string]string{"status": p.Text})
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, m.instance+"/api/v1/statuses", bytes.NewReader(buf))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+m.token)
	res, err := m.http.Do(req)
	if err != nil {
		return "", err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(res.Body, 4096))
		return "", fmt.Errorf("mastodon: status %d: %s", res.StatusCode, bytes.TrimSpace(body))
	}
	var out struct {
		URL string `json:"url"`
	}
	if err := json.NewDecoder(res.Body).Decode(&out); err != nil {
		return "", err
	}
	return out.URL, nil
}

File: apps/ramielle/src/lib/publishers/mastodon.ts
/**
 * Adapter Mastodon (Task 2 da fatia de distribuição) — porte 1:1 de
 * `apps/api/internal/distribution/mastodon.go`. `POST /api/v1/statuses`,
 * header `Authorization: Bearer`, `kind: 'social_hook'`.
 *
 * ⚠️ **Contagem de caracteres é CODE POINT, não `.length`**
 * (`mastodon.go:42`, `utf8.RuneCountInString`), aplicada DEPOIS de anexar o
 * `canonicalUrl` (quando presente) — mesma armadilha do Bluesky, ver o
 * comentário de `bluesky.ts`. `[...texto].length` itera por code point.
 */
import type { DistributionKind } from '../../domain/distribution'
import type { Payload } from './types'

export type MastodonConfig = {
  instanceUrl: string
  token: string
}

const TIMEOUT_MS = 30_000
const MAX_CHARS = 500

export const MASTODON_PLATFORM = 'mastodon'
export const MASTODON_KIND: DistributionKind = 'social_hook'

type MastodonApiResponse = {
  url?: string
}

/**
 * Posta um status no Mastodon. Rejeita textos > 500 code points (após
 * anexar `payload.canonicalUrl`, quando presente). Retorna a URL pública do
 * toot — porte de `Mastodon.Publish` (`mastodon.go:37-68`).
 */
export async function publishMastodon(
  cfg: MastodonConfig,
  payload: Payload,
): Promise<string> {
  let status = payload.text ?? ''
  if (payload.canonicalUrl) {
    status += `\n\n${payload.canonicalUrl}`
  }
  // ⚠️ Code point, não .length — ver o aviso do topo do arquivo.
  if ([...status].length > MAX_CHARS) {
    throw new Error('mastodon: texto excede 500 caracteres')
  }

  // Mesma normalização de `NewMastodon` (`mastodon.go:23-29`, `TrimRight`).
  const instance = cfg.instanceUrl.replace(/\/+$/, '')

  const controller = new AbortController()
  const timeoutId = setTimeout(() => controller.abort(), TIMEOUT_MS)
  let res: Response
  try {
    res = await fetch(`${instance}/api/v1/statuses`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${cfg.token}`,
      },
      body: JSON.stringify({ status }),
      signal: controller.signal,
    })
  } catch {
    throw new Error('mastodon: falha ao executar a requisição de publicação')
  } finally {
    clearTimeout(timeoutId)
  }

  if (res.status < 200 || res.status >= 300) {
    const corpo = await res.text().catch(() => '')
    throw new Error(
      `mastodon: status ${res.status}: ${corpo.slice(0, 4096).trim()}`,
    )
  }

  let out: MastodonApiResponse
  try {
    out = (await res.json()) as MastodonApiResponse
  } catch {
    throw new Error('mastodon: resposta não é um JSON válido')
  }
  return out.url ?? ''
}
